tpu-client-next/src/client_builder.rs

```diff
@@ -0,0 +1,317 @@
+use {
+    crate::{
+        connection_workers_scheduler::{
+            BindTarget, ConnectionWorkersSchedulerConfig, Fanout, StakeIdentity, WorkersBroadcaster,
+        },
+        leader_updater::LeaderUpdater,
+        transaction_batch::TransactionBatch,
+        ConnectionWorkersScheduler, SendTransactionStats,
+    },
+    solana_keypair::Keypair,
+    std::{
+        future::Future,
+        net::{SocketAddr, UdpSocket},
+        pin::Pin,
+        sync::Arc,
+    },
+    thiserror::Error,
+    tokio::{
+        runtime,
+        sync::{mpsc, watch},
+    },
+    tokio_util::{sync::CancellationToken, task::TaskTracker},
+};
+
+/// [`TransactionSender`] provides an interface to send transactions in batches.
+#[derive(Clone)]
+pub struct TransactionSender {
+    sender: mpsc::Sender<TransactionBatch>,
+}
+
+/// [`Client`] runs the background tasks required for sending transactions and
+/// update certificate used the endpoint.
+#[derive(Clone)]
+pub struct Client {
+    update_certificate_sender: watch::Sender<Option<StakeIdentity>>,
+    tasks: TaskTracker,
+    cancel: CancellationToken,
+}
+
+/// [`ClientBuilder`] is a builder structure to create [`TransactionSender`]
+/// along with [`Client`].
+pub struct ClientBuilder {
+    runtime_handle: Option<runtime::Handle>,
+    leader_updater: Box<dyn LeaderUpdater>,
+    bind_target: Option<BindTarget>,
+    identity: Option<StakeIdentity>,
+    num_connections: usize,
+    leader_send_fanout: usize,
+    skip_check_transaction_age: bool,
+    sender_capacity: usize,
+    worker_channel_size: usize,
+    max_reconnect_attempts: usize,
+    report_fn: Option<ReportFn>,
+    cancel: CancellationToken,
+}
+
+impl ClientBuilder {
+    pub fn with_leader_updater(leader_updater: Box<dyn LeaderUpdater>) -> Self {
+        Self {
+            runtime_handle: None,
+            leader_updater,
+            bind_target: None,
+            identity: None,
+            num_connections: 64,
+            leader_send_fanout: 2,
+            skip_check_transaction_age: true,
+            worker_channel_size: 2,
+            sender_capacity: 64,
+            max_reconnect_attempts: 2,
+            report_fn: None,
+            cancel: CancellationToken::new(),
+        }
+    }
+
+    /// Set the runtime handle for the client. If not set, the current runtime
+    /// will be used.
+    ///
+    /// Note that if the runtime handle is not set, the caller must ensure that
+    /// the `build` is called in tokio runtime context. Otherwise, `build` will
+    /// panic.
+    pub fn runtime_handle(mut self, handle: runtime::Handle) -> Self {
+        self.runtime_handle = Some(handle);
+        self
+    }
+
+    /// Set the UDP socket to bind to.
+    pub fn bind_socket(mut self, bind_socket: UdpSocket) -> Self {
+        self.bind_target = Some(BindTarget::Socket(bind_socket));
+        self
+    }
+
+    /// Set the socket address to bind to.
+    pub fn bind_addr(mut self, bind_addr: SocketAddr) -> Self {
```

Comment:
What is the benefit of having this method if we can always set the UDP socket directly?